### PR TITLE
Created 3Party, added gtest and cereal.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi Alex @deathbaba,

Creating `3Party`. Please take a look at:
1. Licenses and the directory structure. I did my best to follow what we have discussed.
2. The way I added gtest (original code https://github.com/dkorolev/3Party/commit/ce2baa9420ce74191da9911622c3e058d563ac56, followed by my port https://github.com/dkorolev/3Party/commit/95ddb4b2770f998a17456a60df05376b838f9b13).
3. Whether it builds and runs on Mac.

Here's how it looks to me now:

```
dima@dima-i9:~/github/dkorolev/3Party/example/gtest$ make
mkdir -p build
g++ -std=c++11 -Wall -o build/demo demo.cc -pthread
./build/demo
[==========] Running 6 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 4 tests from GTestTest
[ RUN      ] GTestTest.Trivial
[       OK ] GTestTest.Trivial (0 ms)
[ RUN      ] GTestTest.Passing
[       OK ] GTestTest.Passing (0 ms)
[ RUN      ] GTestTest.DeathTest
[       OK ] GTestTest.DeathTest (2 ms)
[ RUN      ] GTestTest.ExceptionThrowingTest
[       OK ] GTestTest.ExceptionThrowingTest (0 ms)
[----------] 4 tests from GTestTest (3 ms total)

[----------] 1 test from GTestTemplatedTest/0, where TypeParam = std::vector<int, std::allocator<int> >
[ RUN      ] GTestTemplatedTest/0.TemplatedContainersTest
[       OK ] GTestTemplatedTest/0.TemplatedContainersTest (0 ms)
[----------] 1 test from GTestTemplatedTest/0 (0 ms total)

[----------] 1 test from GTestTemplatedTest/1, where TypeParam = std::deque<int, std::allocator<int> >
[ RUN      ] GTestTemplatedTest/1.TemplatedContainersTest
[       OK ] GTestTemplatedTest/1.TemplatedContainersTest (0 ms)
[----------] 1 test from GTestTemplatedTest/1 (0 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 3 test cases ran. (3 ms total)
[  PASSED  ] 6 tests.

  YOU HAVE 1 DISABLED TEST

dima@dima-i9:~/github/dkorolev/3Party/example/gtest$ 
```

Thanks,
Dima

### CODE REVIEW

**Comment by biodranik** [comment]
AlexMac:gtest alexz$ pwd
/Users/alexz/Developer/projects/3Party/example/gtest
AlexMac:gtest alexz$ make
g++ -std=c++11 -Wall -o build/demo demo.cc -pthread
In file included from demo.cc:27:
In file included from ./../../src/gtest/gtest.h:6:
In file included from ./../../src/gtest/gtest-all.cc.h:42:
./../../src/gtest/./gtest.cc.h:34:10: fatal error: 'gtest/gtest.h' file not found
#include "gtest/gtest.h"
         ^
1 error generated.
make: **\* [build/demo] Error 1

please fix include paths

**Comment by biodranik** [comment]
in short: includes should be similar everywhere. if you use include "gtest/gtest.h" then do not use include "../../gtest/gtest.h" but add -I../../src instead

**Comment by dkorolev** [comment]
Hey, wait. Line 42 of `gtest-all.cc.h` reads `#include "./gtest.cc.h"`. Could you please confirm you're compiling the right gtest?

Perhaps, since it's installed on your system, you need `-I .` ? Worth investigating.

**Comment by dkorolev** [comment]
And let's align on this: by "header-only" and thus "no external dependencies", I mean just #include the header and you're done.

IMHO, making the user to have to do any `-I foo/` ruins most of the purpose of header-only. If the user has to do anything beyond #include-ing the header in order to have the thing work, they might as well link against a library or other nasty things.

I suggest we try to keep our _header-only_ as clean as possible. What do you think?

File: example/gtest/demo.cc [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// A tree-minutes intro to gtest.
```
Comment (biodranik) [inline example/gtest/demo.cc]: three ?

**Comment by biodranik** [comment]
Agree. But not sure that we can make tests header only

**Comment by dkorolev** [comment]
Fixed `tree` to be `three`.

The `gtest` library is now header-only. Each test suit is its own `.cc` and thus an individual binary now, I suggest we keep it this way.

Thanks,
Dima

**Pull request merged**